docs/specs/fee-calculator.md

```diff
@@ -0,0 +1,358 @@
+
+## 2. Purpose
+
+### Primary Objectives
+
+- Calculate daily totals for each tracked reward distributor
+- Analyze balance changes and distribution events to determine total collection
+- Generate comprehensive reports showing daily total breakdowns
+- Support incremental processing for efficient updates
+
+### Component Responsibilities
+
+- Read distributor data, balances, and events through FileManager
+- Compute daily total amounts using balance changes and distributions
+- Validate data consistency before calculations
+- Output structured fee reports
+
+### Success Criteria
+
+- All distributors have accurate daily total calculations for applicable dates
+- Total calculations correctly reflect the formula: Daily Total = Balance Change + Distributions
+- Component throws error if critical data is missing
+- Reports clearly show total breakdowns per distributor per day
+
+## 3. Dependencies
+
+### External Libraries
+
+- Node.js built-in modules: `fs/promises`, `path`
+- No external dependencies required
+
+### System Components
+
+- **FileManager**: Exclusive interface for all data access (distributors, balances, events)
+
+### Infrastructure Requirements
+
+- Read access to stored distributor, balance, and event data through FileManager
+- Write permissions for total calculation results
+
+## 4. Data Flow
+
+### Input Flow
+
+All data access exclusively through FileManager functions:
+
+1. Use FileManager to read list of all tracked distributors
+2. For each distributor:
+   - Use FileManager to read balance history data
+   - Use FileManager to read distribution event data
+
+### Process Flow
+
+1. Load all distributor addresses from FileManager
+2. For each distributor:
+   - Retrieve complete balance history
+   - Retrieve all distribution events
+   - Sort data chronologically
+   - For each day with balance data:
+     - Calculate balance change from previous day
+     - Sum distributions for that day
+     - Compute daily total = balance change + distributions
+   - Store calculated totals in memory
+3. Generate comprehensive report
+4. Write results using FileManager
+
+### Output Flow
+
+- Total calculation results written through FileManager
+- Single report file containing all distributor daily totals (completely overwrites any existing report)
+- Structured format supporting easy querying and analysis
+
+## 5. Public API
+
+### Main Entry Point
+
+```typescript
+interface FeeCalculator {
+  calculateFees(distributorAddress?: string): Promise<void>;
+}
+```
+
+### Method Signatures
+
+```typescript
+calculateFees(distributorAddress?: string): Promise<void>
+```
+
+- **Purpose**: Calculate daily totals for all distributors or a specific distributor
+- **Parameters**:
+  - `distributorAddress` (optional): If provided, only calculate fees for this distributor
+- **Returns**: Promise that resolves when total calculations are complete
```

Comment:
No, it shouldn't return a promise and we don't want a promise for future extensibility. This is a synchronous method. We don't call any async methods from within.